client/src/txnmgr.rs

```diff
@@ -0,0 +1,326 @@
+
+        let keys: Vec<Txid> = txn_map.keys().cloned().into_iter().collect();
+
+        let result = event::wait(
+            &events,
+            |e| match e {
+                event::Event::Received(e, NetworkMessage::Inv(inv)) => {
+                    let mut found_requested_tx = false;
+
+                    for key in &keys {
+                        if inv.contains(&Inventory::Transaction(*key)) {
+                            let (txn, peers_recvd, _) = txn_map.get_mut(key).unwrap();
+
+                            let result = self
+                                .handle
+                                .command(Command::SubmitTransaction(e, txn.to_owned()));
+
+                            peers_recvd.insert(e);
+
+                            match result {
+                                Ok(_) => {
+                                    // We continue tracking for transactions in store even when
+                                    // requested tx is found, so we dont miss any event.
+                                    if *key == tx_id {
+                                        found_requested_tx = true;
+                                    }
+                                }
+
+                                Err(e) => return Some(Err(e)),
+                            }
+                        }
+                    }
+
+                    if found_requested_tx {
+                        return Some(Ok(()));
+                    }
+
+                    None
+                }
+
+                event::Event::SyncManager(syncmgr::Event::BlockReceived(_, block, height)) => {
+                    self.on_blocks_received(tx_id, block, height, &mut *txn_map)
+                }
+
+                _ => None,
+            },
+            timeout,
+        );
+
+        let (_, peers_recvd, height) = txn_map.get_mut(&tx_id).unwrap();
+
+        // If the first unwrapped error is sent, it signifies event
+        // timed out. The second error value is that when tracking transaction.
+        match result? {
+            Ok(_) => {
+                if *height == 0 {
+                    Ok(Event::Accepted {
+                        confirmations: peers_recvd.len(),
+                        txid: tx_id,
+                    })
+                } else {
+                    Ok(Event::Confirmed {
+                        height: *height,
+                        txid: tx_id,
+                    })
+                }
+            }
+
+            Err(e) => Err(e),
+        }
+    }
+
+    fn on_blocks_received(
+        &self,
+        tx_id: Txid,
+        block: Block,
+        height: Height,
+        txn_store: &mut HashMap<Txid, (Transaction, HashSet<PeerId>, Height)>,
+    ) -> Option<Result<(), handle::Error>> {
+        let mut found_requested_tx = false;
+
+        for (id, (txn, _, confirmed_height)) in txn_store.iter_mut() {
+            if block.txdata.contains(txn) {
+                *confirmed_height = height;
```

Comment:
It would make more sense if the txn_store was just the Mempool. Otherwise, what is the confirmed height of an unconfirmed tx?

Reply:
We are indicating in the storage that the specific transaction has been confirmed at the specified height. This is to ensure that even even when users request for a transaction status right after we've sent a `TransactionAccepted` status, we are able to still indicate the transaction has been confirmed.